inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/special/ClassLoaderDelegation.java

```diff
@@ -1,165 +1,103 @@
-     */
-    public boolean wasBootstrapMadeAvailableTo(ClassLoader classLoader) {
-        if (classLoader == null) {
-            return true;
-        }
-        Boolean isBootstrapAccessible = bootstrapClassesAccessible.getIfPresent(classLoader);
-        return isBootstrapAccessible != null && isBootstrapAccessible;
+    private final Set<Class<?>> instrumentedClassloaderClasses = Collections.newSetFromMap(
+            CacheBuilder.newBuilder().weakKeys().<Class<?>, Boolean>build().asMap());
+
+    private final ElementMatcher<MethodDescription> LOAD_CLASS_MATCHER =
+            named("loadClass").and(
+                    takesArguments(String.class, boolean.class)
+                            .or(takesArguments(String.class)));
+
+    private final ElementMatcher<TypeDescription> CLASS_LOADER_MATCHER =
+            isSubTypeOf(ClassLoader.class)
+                    .and(not(nameStartsWith("java.")))
+                    .and(declaresMethod(LOAD_CLASS_MATCHER));
+
+    @Override
+    public boolean shouldInstrument(Class<?> clazz, InstrumentationConfiguration settings) {
+        boolean isClassLoader = CLASS_LOADER_MATCHER.matches(TypeDescription.ForLoadedType.of(clazz));
+        boolean cldEnabled = settings.getSource().getSpecial().isClassLoaderDelegation();
+        //we never de instrument a classloader we previously instrumented
+        return instrumentedClassloaderClasses.contains(clazz) || (isClassLoader && cldEnabled);
     }
 
-    /**
-     * Tries to make the inspectIT bootstrap classes such as {@link Instances} available to the given classloader.
-     * This method first checks if the classes are available by default.
-     * If this is not the case, it tries to instrument the classloader with the classloader delegation via {@link ClassloaderDelegationSensor}.
-     * <p>
-     * The result of this method invocation is cached, meaning that repeated calls cause no overhead.
-     *
-     * @param classLoader the classloader through which our bootstrap classes should be available.
-     * @return true, if the bootstrap classes are now available to the given classloader. False otherwise, in
-     * which case no instrumentation should be applied to any class loaded by this classloader.
-     */
-    public boolean makeBootstrapClassesAvailable(ClassLoader classLoader) {
-        if (classLoader == null) {
-            return true;
-        }
-
-        Boolean isBootstrapAccessible = bootstrapClassesAccessible.getIfPresent(classLoader);
-        if (isBootstrapAccessible == null) {
-            return makeBootstrapClassesAvailableSynchronized(classLoader);
-        } else {
-            return isBootstrapAccessible;
-        }
+    @Override
+    public boolean requiresInstrumentationChange(Class<?> clazz, InstrumentationConfiguration first, InstrumentationConfiguration second) {
+        return false;
```

Comment:
This remaining false is still correct? We will not react to the configuration updates?

Reply:
Yes, we do react via shouldInstrument, this method only tests if changes are necessary while the sensor remained active.